### PR TITLE
[ZEPPELIN-1666] DON'T share mutable deps, options between interpreters in each group (bug)

### PR DESCRIPTION
**Pull request by 1ambda**
### What is this PR for?

Every interpreter shares their `List<Dependency>` and `InterpreterOption` object with other interpreters in the same group since these objects are mutable and just returned from `InterpreterSettingRef` in InterpreterFactory.

I attached GIF

### What type of PR is it?
[Bug Fix]

### Todos

Nothing

### What is the Jira issue?

[ZEPPELIN-1666](https://issues.apache.org/jira/browse/ZEPPELIN-1666)

### How should this be tested?

I included unit test for it in `InterpreterRestApiTest.testCreatedInterpreterDependencies`.   
You can reproduce and debug this by checking out commit 3acde56 and run the unit test.

If you try to debug, you can see the length of `List<Dependency>` in `md` `InterpreterSettingRef` is increased whenever creating a new interpreter setting. But it shouldn't be.

### Screenshots (if appropriate)

duplicated_interpreter_params

### Questions:
* Does the licenses files need update? - NO
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - NO

### CODE REVIEW

**Comment by 1ambda** [comment]
You can test them by checking out commit 74773e572b44a31b3082cd082aa4a097de9d0e63 in this PR and run `InterpreterRestApiTest.testCreatedInterpreterDependencies`

**Comment by 1ambda** [comment]
Build profiles 2, 3, 4, 8 are nondeterministic :(

**Comment by 1ambda** [comment]
Currently all interpreters in each group shares same dependencies and options. So if we create 2 interpreters `postgres1` `postgres2` and set a dependency `postgres94.client.jar` `postgres95.client.jar` seperately then they have same 2 dependencies. (**bug**)

I fixed this issue but this results in a lot of copying dependencies in CI test. So it eventually failed while throwing this message

```
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.19 sec - in org.apache.zeppelin.rest.NotebookSecurityRestApiTest
Running org.apache.zeppelin.server.CorsFilterTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.175 sec - in org.apache.zeppelin.server.CorsFilterTest
Running org.apache.zeppelin.ticket.TicketContainerTest
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec - in org.apache.zeppelin.ticket.TicketContainerTest
Running org.apache.zeppelin.security.SecurityUtilsTest


The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).

The job has been terminated
```

Test was passing but finally failed because of the log size. Could anyone receommend an idea to resolve this? (I think Travis CI will not increase log size https://github.com/travis-ci/travis-ci/issues/1382)

**Comment by 1ambda** [comment]
```
05:33:43,422  INFO alluxio.logger.type:202 - Connecting to local worker @ testing-docker-d2b67f49-f525-4c23-a9ab-36aafddcf032/172.17.0.12:42241
05:33:43,422 ERROR alluxio.logger.type:212 - java.net.ConnectException: Connection refused
org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused
	at org.apache.thrift.transport.TSocket.open(TSocket.java:187)
	at org.apache.thrift.transport.TFramedTransport.open(TFramedTransport.java:81)
	at alluxio.client.block.BlockWorkerClient.connectOperation(BlockWorkerClient.java:210)
	at alluxio.client.block.BlockWorkerClient.connect(BlockWorkerClient.java:304)
	at alluxio.AbstractClient.retryRPC(AbstractClient.java:291)
	at alluxio.client.block.BlockWorkerClient.sessionHeartbeat(BlockWorkerClient.java:408)
	at alluxio.client.block.BlockWorkerClient.periodicHeartbeat(BlockWorkerClient.java:424)
	at alluxio.client.block.BlockWorkerClientHeartbeatExecutor.heartbeat(BlockWorkerClientHeartbeatExecutor.java:34)
	at alluxio.heartbeat.HeartbeatThread.run(HeartbeatThread.java:67)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:339)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:200)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:182)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:579)
	at org.apache.thrift.transport.TSocket.open(TSocket.java:182)
	... 13 more
```

```
05:33:43,416  INFO alluxio.logger.type:151 - Folder /tmp/Alluxio871779581980018644U1479706421593/ramdisk/datastore/2965139340539231725 was created!
05:33:43,417  INFO alluxio.logger.type:65 - LocalBlockOutStream created new file block, block path: /tmp/Alluxio871779581980018644U1479706421593/ramdisk/datastore/2965139340539231725/16777389
05:33:43,421 ERROR alluxio.logger.type:297 - 
org.apache.thrift.transport.TTransportException
	at org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:132)
	at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at org.apache.thrift.transport.TFramedTransport.readFrame(TFramedTransport.java:129)
	at org.apache.thrift.transport.TFramedTransport.read(TFramedTransport.java:101)
	at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:429)
	at org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:318)
	at org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:219)
	at org.apache.thrift.protocol.TProtocolDecorator.readMessageBegin(TProtocolDecorator.java:135)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:69)
	at alluxio.thrift.BlockWorkerClientService$Client.recv_sessionHeartbeat(BlockWorkerClientService.java:424)
	at alluxio.thrift.BlockWorkerClientService$Client.sessionHeartbeat(BlockWorkerClientService.java:410)
	at alluxio.client.block.BlockWorkerClient$10.call(BlockWorkerClient.java:411)
	at alluxio.client.block.BlockWorkerClient$10.call(BlockWorkerClient.java:408)
	at alluxio.AbstractClient.retryRPC(AbstractClient.java:293)
	at alluxio.client.block.BlockWorkerClient.sessionHeartbeat(BlockWorkerClient.java:408)
	at alluxio.client.block.BlockWorkerClient.periodicHeartbeat(BlockWorkerClient.java:424)
	at alluxio.client.block.BlockWorkerClientHeartbeatExecutor.heartbeat(BlockWorkerClientHeartbeatExecutor.java:34)
	at alluxio.heartbeat.HeartbeatThread.run(HeartbeatThread.java:67)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

```
ERROR [04:32:57,995][] org.apache.zeppelin.cassandra.InterpreterLogic@:interpret All host(s) tried for query failed (tried: /127.0.0.1:9407 (com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency THREE (3 required but only 1 alive))) 
com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /127.0.0.1:9407 (com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency THREE (3 required but only 1 alive)))
	at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:84)
	at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:37)
	at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
	at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:245)
	at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:63)
	at org.apache.zeppelin.cassandra.EnhancedSession.execute(EnhancedSession.scala:200)
	at org.apache.zeppelin.cassandra.InterpreterLogic$$anonfun$8.apply(InterpreterLogic.scala:155)
	at org.apache.zeppelin.cassandra.InterpreterLogic$$anonfun$8.apply(InterpreterLogic.scala:155)
	at scala.collection.immutable.List.map(List.scala:273)
	at org.apache.zeppelin.cassandra.InterpreterLogic.interpret(InterpreterLogic.scala:155)
	at org.apache.zeppelin.cassandra.CassandraInterpreter.interpret(CassandraInterpreter.java:200)
	at org.apache.zeppelin.cassandra.CassandraInterpreterTest.should_execute_statement_with_consistency_option(CassandraInterpreterTest.java:236)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
	at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /127.0.0.1:9407 (com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency THREE (3 required but only 1 alive)))
	at com.datastax.driver.core.RequestHandler.reportNoMoreHosts(RequestHandler.java:211)
	at com.datastax.driver.core.RequestHandler.access$1000(RequestHandler.java:43)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.sendRequest(RequestHandler.java:277)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution$1.run(RequestHandler.java:400)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Tests run: 35, Failures: 0, Errors: 0, Skipped: 3, Time elapsed: 54.761 sec - in org.apache.zeppelin.cassandra.CassandraInterpreterTest
```

```
ERROR [04:32:55,197][] org.apache.zeppelin.cassandra.InterpreterLogic@:interpret The statement 'select_users' can not be bound to values. Are you sure you did prepare it with @prepare[select_users] ? 
org.apache.zeppelin.interpreter.InterpreterException: The statement 'select_users' can not be bound to values. Are you sure you did prepare it with @prepare[select_users] ?
	at org.apache.zeppelin.cassandra.InterpreterLogic.generateBoundStatement(InterpreterLogic.scala:303)
	at org.apache.zeppelin.cassandra.InterpreterLogic$$anonfun$6.apply(InterpreterLogic.scala:149)
	at org.apache.zeppelin.cassandra.InterpreterLogic$$anonfun$6.apply(InterpreterLogic.scala:139)
	at scala.collection.immutable.List.map(List.scala:273)
	at org.apache.zeppelin.cassandra.InterpreterLogic.interpret(InterpreterLogic.scala:139)
	at org.apache.zeppelin.cassandra.CassandraInterpreter.interpret(CassandraInterpreter.java:200)
	at org.apache.zeppelin.cassandra.CassandraInterpreterTest.should_exception_when_executing_unknown_bound_statement(CassandraInterpreterTest.java:358)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
	at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

**Comment by astroshim** [comment]
Thank you for fixing bug.
It's working as expected and adding test is also great!

**Comment by 1ambda** [comment]
I will rebase to clean useless commits

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java [hunk 1]
```diff
@@ -133,20 +134,23 @@
       "editOnDblClick", false);
 
   public InterpreterFactory(ZeppelinConfiguration conf,
-      AngularObjectRegistryListener angularObjectRegistryListener,
-      RemoteInterpreterProcessListener remoteInterpreterProcessListener,
-      ApplicationEventListener appEventListener, DependencyResolver depResolver,
-      boolean shiroEnabled) throws InterpreterException, IOException, RepositoryException {
+                            AngularObjectRegistryListener angularObjectRegistryListener,
+                            RemoteInterpreterProcessListener remoteInterpreterProcessListener,
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java]: Could you please make sure that all the code changes adhere Zeppelin's [project style guide](https://zeppelin.apache.org/contribution/contributions.html#code-convention)?

In particular, [horizontal code alignment is discouraged](https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment) by project's Java language style guide due to mainanability concerns. 

Could you please change it back here and everywhere below?

Ther checkstyle right now verifies only sub-set of styleguide preferences and there is on-going (but stalled) effort to improve it, introducing appropriate consistent cross-IDE formating in #1168

**Comment by bzz** [comment]
Looks great to me, except for code style noted above.

Thank you @1ambda for tests improvements, would be happy to merge as soon as the style issue addressed, if there is no further discussion.

**Comment by bzz** [comment]
@1ambda ping. Would be great to merge it ASAP if it helps \w CI stability

**Comment by 1ambda** [comment]
@bzz Removed styling commit. Let's see the CI result. Thanks!

**Comment by bzz** [comment]
Awesome, 👍  for tests and a great commit\history structure that allows just to skip one and preserve all the work!

**Comment by 1ambda** [comment]
Something is wrong. I will fix it.

**Comment by bzz** [comment]
Great, CI is green now!

Merging to master, if there is no further discussion

**Pull request closed without merging**